File: basic_java/src/main/java/pers/xiaoming/spark/core/HeapAccumulator.java
package pers.xiaoming.spark.core;

import org.apache.spark.util.AccumulatorV2;

import java.util.ArrayList;
import java.util.List;
import java.util.PriorityQueue;

class HeapAccumulator extends AccumulatorV2<Integer, List<Integer>> {
    private final PriorityQueue<Integer> heap;
    private final int n;

    public HeapAccumulator(int n) {
        this.n = n;
        this.heap = new PriorityQueue<>(n);
    }

    public HeapAccumulator(PriorityQueue<Integer> heap, int n) {
        this.heap = heap;
        this.n = n;
    }

    @Override
    public boolean isZero() {
        return heap.isEmpty();
    }

    @Override
    public AccumulatorV2<Integer, List<Integer>> copy() {
        return new HeapAccumulator(this.heap, this.n);
    }

    @Override
    public void reset() {
        heap.clear();
    }

    @Override
    public void add(Integer v) {
        if (heap.size() < n) {
            heap.add(v);
        } else {
            if (heap.peek() < v) {
                heap.poll();
                heap.add(v);
            }
        }
    }

    @Override
    public void merge(AccumulatorV2<Integer, List<Integer>> other) {
        for (int v : other.value()) {
            this.add(v);
        }
    }

    @Override
    public List<Integer> value() {
        return new ArrayList<>(heap);
    }
}

File: basic_scala/src/main/scala/pers/xiaoming/spark/core/HeapAccumulator.scala
package pers.xiaoming.spark.core

import Ordering.Implicits._
import org.apache.spark.util.AccumulatorV2

import scala.collection.mutable
import scala.reflect.ClassTag

class HeapAccumulator[T : Ordering : ClassTag](private val n:Int, private val heap:mutable.PriorityQueue[T])
  extends AccumulatorV2[T, List[T]] {

  def this(n:Int) = this(n, mutable.PriorityQueue.empty[T](Ordering[T].reverse))

  override def isZero: Boolean = heap.isEmpty

  override def copy(): AccumulatorV2[T, List[T]] = new HeapAccumulator[T](n, heap)

  override def reset(): Unit = heap.clear()

  override def add(v: T): Unit = {
    if (heap.size < n) {
      heap.+=(v)
    } else {
      if (heap.head < v) {
        heap.dequeue()
        heap.+=(v)
      }
    }
  }

  override def merge(other: AccumulatorV2[T, List[T]]): Unit = {
    other.value.foreach(heap.+=)
  }

  override def value: List[T] = heap.toList
}
